apis/projectcontour/v1/helpers.go

```diff
@@ -78,3 +83,211 @@ func (r *Route) AuthorizationContext(parent map[string]string) map[string]string
 
 	return values
 }
+
+// AddError adds an error-level Subcondition to the DetailedCondition.
+// AddError will also update the DetailedCondition's state to take into account
+// the error that's present.
+// If a SubCondition with the given errorType exists, will change
+// the reason to "MultipleReasons", and will append the given message
+// to the existing SubCondition's message.
+func (dc *DetailedCondition) AddError(errorType, reason, message string) {
+	message = truncateLongMessage(message)
+
+	// Update the condition so that it indicates there's at least one error
+	// This needs to be here because conditions may be normal-true (positive)
+	// polarity (like `Valid`), or abnormal-true (negative) polarity
+	// (like `ErrorPresent`)
+	if dc.IsPositivePolarity() {
+		dc.Status = ConditionFalse
+	} else {
+		dc.Status = ConditionTrue
+	}
+
+	// Update the DetailedCondition Reason and message if the reason
+	// is different to the current one. We have to namespace
+	// the DetailedCondition reason by the errorType because we're
+	// adding SubConditions; DetailedCondition Reason is a summary of the
+	// SubCondition reasons.
+	detailedReason := errorType + reason
+	dc.updateReason(detailedReason, message)
+	// Now we can actually upsert the error.
+	i := getIndex(errorType, dc.Errors)
+
+	if i == -1 {
+		dc.Errors = append(dc.Errors, SubCondition{
+			Type:    errorType,
+			Status:  ConditionTrue,
+			Message: message,
+			Reason:  reason,
+		})
+		return
+	}
+
+	// If the error is already present, just update the reason and message.
+	dc.Errors[i].updateReason(reason, message)
+
+}
+
+// AddErrorf adds an error-level Subcondition to the DetailedCondition, using
+// fmt.Sprintf on the formatmsg and args params.
+// If a SubCondition with the given errorType exists, will change
+// the reason to "MultipleReasons", and will append the given message
+// to the existing SubCondition's message.
+func (dc *DetailedCondition) AddErrorf(errorType, reason, formatmsg string, args ...interface{}) {
+	dc.AddError(errorType, reason, fmt.Sprintf(formatmsg, args...))
+}
+
+// GetError gets an error of the given errorType.
+// Similar to a hash lookup, will return true in the second value if a match is
+// found, and false otherwise.
+func (dc *DetailedCondition) GetError(errorType string) (SubCondition, bool) {
+	i := getIndex(errorType, dc.Errors)
+
+	if i == -1 {
+		return SubCondition{}, false
+	}
+
+	return dc.Errors[i], true
+}
+
+// AddWarning adds an warning-level Subcondition to the DetailedCondition.
+// If a SubCondition with the given warnType exists, will change
+// the reason to "MultipleReasons", and will append the given message
+// to the existing SubCondition's message.
+func (dc *DetailedCondition) AddWarning(warnType, reason, message string) {
+	message = truncateLongMessage(message)
+
+	detailedReason := warnType + reason
+	dc.updateReason(detailedReason, message)
```

Comment:
Why does adding a warning update the top-level condition? If there are warnings, then I don't think that we would rewrite the top-level reason (at least in the case of the Valid condition).

Reply:
Fine.